File: golang/2106.maximum-fruits-harvested-after-at-most-k-steps.go
type Positions []int

func (p *Positions) bisectLeft(x int) int {
	arr := *p

	lo, hi := 0, len(arr)
	var mid int
	for lo < hi {
		mid = (lo + hi) / 2

		if arr[mid] < x {
			lo = mid + 1
		} else {
			hi = mid
		}
	}
	return lo
}

func (p *Positions) bisectRight(x int) int {
	arr := *p

	lo, hi := 0, len(arr)
	var mid int

	for lo < hi {
		mid = (lo + hi) / 2

		if arr[mid] > x {
			hi = mid
		} else {
			lo = mid + 1
		}
	}

	return lo
}

func maxTotalFruitsV1(fruits [][]int, startPos int, k int) int {
	n := len(fruits)
	pos := make(Positions, n)
	prefix := make([]int, n+1)

	for i := range n {
		pos[i] = fruits[i][0]
		prefix[i+1] = prefix[i] + fruits[i][1]
	}

	var left, right, start, end, curMax int

	for x := range k/2 + 1 {
		left = startPos - x
		right = startPos + k - 2*x
		start = pos.bisectLeft(left)
		end = pos.bisectRight(right)
		curMax = max(curMax, prefix[end]-prefix[start])

		right = startPos + x
		left = startPos - k + 2*x
		start = pos.bisectLeft(left)
		end = pos.bisectRight(right)
		curMax = max(curMax, prefix[end]-prefix[start])
	}

	return curMax
}

func countSteps(fruits [][]int, startPos, left, right int) int {
	leftPart := startPos - fruits[left][0]
	rightPart := fruits[right][0] - startPos
	both := leftPart + rightPart

	if leftPart <= 0 {
		return rightPart
	} else if rightPart <= 0 {
		return leftPart
	}

	return min(leftPart, rightPart) + both
}

func maxTotalFruitsV2(fruits [][]int, startPos int, k int) int {
	n := len(fruits)
	cur, curMax := 0, 0

	for left, right := 0, 0; right < n; right++ {
		cur += fruits[right][1]

		for left <= right && countSteps(fruits, startPos, left, right) > k {
			cur -= fruits[left][1]
			left++
		}

		curMax = max(curMax, cur)
	}

	return curMax
}

func maxTotalFruits(fruits [][]int, startPos int, k int) int {
	return maxTotalFruitsV2(fruits, startPos, k)
}

File: cpp/2106.maximum-fruits-harvested-after-at-most-k-steps.cpp
#include <vector>
using std::vector;

class Solution {
private:
  int bisectLeft(const vector<int> &arr, int x) {
    int lo = 0, hi = arr.size();
    int mid;

    while (lo < hi) {
      mid = (lo + hi) / 2;

      if (arr[mid] < x) {
        lo = mid + 1;
      } else {
        hi = mid;
      }
    }

    return lo;
  }

  int bisectRight(const vector<int> &arr, int x) {
    int lo = 0, hi = arr.size();
    int mid;

    while (lo < hi) {
      mid = (lo + hi) / 2;

      if (arr[mid] > x) {
        hi = mid;
      } else {
        lo = mid + 1;
      }
    }

    return lo;
  }

  int maxTotalFruitsV1(vector<vector<int>> &fruits, int startPos, int k) {
    const int n = fruits.size();
    vector<int> pos(n, 0);
    vector<int> prefix(n + 1, 0);

    for (int i = 0; i < n; i++) {
      pos[i] = fruits[i][0];
      prefix[i + 1] = prefix[i] + fruits[i][1];
    }

    int curMax = 0, left, right, start, end;
    for (int x = 0; x <= k / 2; x++) {
      left = startPos - x;
      right = startPos + k - 2 * x;

      start = bisectLeft(pos, left);
      end = bisectRight(pos, right);

      curMax = std::max(curMax, prefix[end] - prefix[start]);

      left = startPos - k + 2 * x;
      right = startPos + x;

      start = bisectLeft(pos, left);
      end = bisectRight(pos, right);

      curMax = std::max(curMax, prefix[end] - prefix[start]);
    }

    return curMax;
  }

  int stepsCount(const vector<vector<int>> &fruits, int startPos, int left,
                 int right) {
    int leftPart = startPos - fruits[left][0];
    int rightPart = fruits[right][0] - startPos;
    int bothParts = rightPart + leftPart;

    if (startPos <= fruits[left][0]) {
      return rightPart;
    } else if (startPos >= fruits[right][0]) {
      return leftPart;
    }

    return std::min(leftPart, rightPart) + bothParts;
  }

  int maxTotalFruitsV2(vector<vector<int>> &fruits, int startPos, int k) {
    const int n = fruits.size();
    int cur = 0, curMax = 0, left = 0;

    for (int right = 0; right < n; right++) {
      cur += fruits[right][1];

      while (left <= right && stepsCount(fruits, startPos, left, right) > k) {
        cur -= fruits[left][1];
        left++;
      }

      curMax = std::max(curMax, cur);
    }
    return curMax;
  }

public:
  int maxTotalFruits(vector<vector<int>> &fruits, int startPos, int k) {
    return maxTotalFruitsV2(fruits, startPos, k);
  }
};